Updating Vue Quickstart for auth0-spa-js (#8217)
auth0/docs: articles/quickstart/spa/vuejs/02-calling-an-api.md

---
title: Calling an API
description: This tutorial demonstrates how to make calls to an external API
budicon: 546
topics:
  - quickstarts
  - spa
  - vuejs
  - apis
github:
  path: 02-Calling-an-API
sample_download_required_data:
  - client
  - api
contentType: tutorial
useCase: quickstart
---

> <!-- markdownlint-disable MD002 MD041 -->

Reviewer: TIL! 🎉

Author: :D

> Most single-page apps use resources from data APIs. You may want to restrict access to those resources so that only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).

Reviewer 2: Most single-page apps use resources from data APIs. You may want to restrict these resources so only authenticated users with sufficient privileges can access them. Auth0 lets you manage access to these resources using [API Authorization](/api-auth).

> In this tutorial, you'll create a simple API using [Express](https://expressjs.com) to validate incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.

Reviewer: Same feedback on "JSON Web Tokens"

> This tutorial shows you how to create a simple API using [Express](https://expressjs.com) that validates incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.

Reviewer 2: In this tutorial, you'll create a simple API using [Express](https://expressjs.com) to validate incoming JSON Web Tokens. You will then see how to call this API using an access token granted by the Auth0 authorization server.

<%= include('../_includes/_calling_api_create_api') %>

<%= include('../_includes/_calling_api_create_backend.md') %>

Finally, modify `package.json` to add two new scripts `dev` and `api` that can be used to start the frontend and the backend API together:

```json
{
  "name": "03-calling-an-api",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "serve": "vue-cli-service serve",
    "build": "vue-cli-service build",
    "lint": "vue-cli-service lint",
    "dev": "npm-run-all --parallel serve api",
    "api": "node server.js"
  },

  // .. package dependencies and other JSON nodes
}
```

You can now start the project using `npm run dev` in the terminal, and the frontend Vue.js application will start up alongside the backend API.

### Set up a proxy to the backend API

> To call the API from the frontend application, the development server must be configured to proxy requests through to the backend API. To do this, add a `vue.config.js` file to the root of the project and populate it with the following code:

Reviewer: Is this essential to the quickstart?

Author: @joshcanhelp It's a tricky one that I have constantly battled with. It's not essential to showing how Auth0 works, no. If we take this out, we'd probably have to take out everything to do with creating the backend server.

I'd be in favour of doing this as it would condense the quickstart down further to just the essentials while still showing how to _call_ an API. We could leave the actual backend in the samples as an example for the reader to view.

The only issue then is that the other SPA quickstarts have this backend. Options:

1) Leave the backend documentation where it is for consistency with other SPA quickstarts
2) Remove it in favour of making the quickstart more concise

Personally I'm leaning towards 2. We could try it and see what kind of feedback we get on it, if any?

Reviewer: Now I'm split :)

My default decision is to remove anything not related to the task at hand. If it does not improve the experience, increase learning, or reduce overall complexity, I don't think it should be there. If, however, this increases the chance of success for some reason (smooth out gaps in a specific platform or something), then that's a positive. 

I'll leave it up to you but it _seems_ like this just makes a nice URL for the API which, in this case, is not relevant for the task at hand.

Author: It's more to make the API accessible on the same origin. When considering the bigger picture, it makes configuration and deployment a bit easier too.

I think it's more a question of whether the docs needs to detail how the sample app is built, or leave it out of the docs and have readers study the sample app for this kind of detail.

For now, I'm going to leave it as it is, as this is detail is also in the other SPA quickstarts and we haven't had any feedback on it to make us think it's detrimental to the reader's experieance. Howver, I do hope that this is something that the current work into quickstart principals can help definitively answer.

Reviewer: > It's more to make the API accessible on the same origin.

Ahhhh, that didn't occur to me. Makes sense to me. Thank you for the thorough explanation.

```js
// vue.config.js

module.exports = {
  devServer: {
    proxy: {
      "/api": {
        target: "http://localhost:3001"
      }
    }
  }
};
```

::: note
This assumes that your project was created using [Vue CLI 3](https://cli.vuejs.org/guide/). If your project was not created in the same way, the above should be included as part of your Webpack configuration.
:::

With this in place, the frontend application can make a request to `/api/external` and it will be correctly proxied through to the backend API at `http://localhost:3001/api/external`.

### Restart the application

If you are already running the application using `npm run serve`, then stop that process for now. Restart the app using the following command, which will run both the frontend app and the backend API together:

> ```bash

Reviewer: Nice little command here!

PORT=3000 npm run dev
```

## Add the Audience

> The API backend you created above expects API calls to have a JSON Web Token for authorization. To do this, you must instruct Auth0 to emit an access token in JWT format. This is done by supplying the API identifier when your users authenticate, telling Auth0 which API you want an access token for.

Reviewer: ```suggestion
The API backend you created above expects API calls to have a JSON Web Token for authorization. To do this, you must instruct Auth0 to issue an access token in JWT format. This is done by supplying the API identifier in an `audience` parameter when users authenticate, telling Auth0 which API you want an access token for.
```

> The API backend that you created above expects API calls to have a JSON Web Token attached so that it can authorize the call. To do this, you must instruct Auth0 to emit an access token in JWT format. This is done by supplying the API identifier when your users authenticate, telling Auth0 which API you want an access token for.

Reviewer 2: The API backend you created above expects API calls to have a JSON Web Token for authorization.

Open `auth_config.json` in the root of the project and make sure that a value for `audience` is exported along with the other settings. The value for `audience` should be the "API identifier" for the API you created above:

```json
{
  "domain": "${account.namespace}",
  "clientId": "${account.clientId}",
  "audience": "${apiIdentifier}"
}
```

Finally, open `main.js` and configure the Auth0 plugin with this audience value:

```js

// .. other imports

// NEW - import the audience

> import { domain, clientId, audience } from "../auth_config.json";

Reviewer: Great way to break the config out!

// ..

Vue.use(Auth0Plugin, {
  domain,
  clientId,
  audience,   // NEW - configure the plugin with the audience value
  onRedirectCallback: appState => {
    router.push(
      appState && appState.targetUrl
        ? appState.targetUrl
        : window.location.pathname
    );
  }
});

// .. more Vue configuration ..

```

## Call the API Using an Access Token

> The frontend Vue.js application should be modified to include a page that calls the API using an access token. Similar to the previous tutorial, this includes modifying the Vue router and adding a new view with a button that calls the API.

Reviewer 2: Modify the frontend Vue.js application to include a page that calls the API using an access token. Like the previous tutorial, you'll need to change the Vue router and add a new view with a button that calls the API.

### Install the Axios package

Install the [`axios`](https://www.npmjs.com/package/axios) HTTP library, which will allow us to make HTTP calls out to the backend API:

```bash
npm install axios
```

:::note
Axios is just used as an example; you can use any JavaScript HTTP client that allows you to specify the `Authorization` header to perform this task
:::

### Add a new page

Next, create a new file `ExternalApi.vue` inside the `views` folder. Add a button to call the API and an HTML element that can display the result of the API call.

Use the `getTokenSilently` method of the Auth0 wrapper to get an access token, and then use Axios to make the API call. Attach the access token to the call by setting it as the value for the `Authorization` header, as in the following example:

```html
<template>
  <div>

>     <button @click="callApi">Ping</button>

Reviewer: ```suggestion
    <button @click="callApi">Call</button>
```

Author: Will change this on the sample too.

    }
  }
};
</script>
```

Modify the Vue router to include a route to this new page whenever the `/external-api` URL is accessed:

```js
// src/router.js

// .. other imports

// NEW - import the view for calling the API
import ExternalApiView from "./views/ExternalApi.vue";

const router = new Router({
  mode: "history",
  base: process.env.BASE_URL,
  routes: [
    // ... other routes,

    // NEW - add a new route for the new page
    {
      path: "/external-api",
      name: "external-api",
      component: ExternalApiView,
      beforeEnter: authGuard
    }
  ]
});
```

Note that this new route also uses `authGuard` from the previous tutorial, which prevents access to the page unless the user is authenticated.

Finally, modify the navigation bar to include a link to the new page:

```html
<template>
  <div id="app">
    <div id="nav">
      <router-link to="/">Home</router-link>|
      <router-link to="/about">About</router-link>|

      <!-- NEW - add a link to the new page -->
      <router-link to="/external-api">External Api</router-link>

      <!-- .. other HTML elements -->

    </div>
    <router-view />
  </div>
</template>
```

:::panel Checkpoint

> Now you will be able to run the application, browse to the "External API" page and press the "Ping" button. The application will make a call to the external API endpoint and produce a message on the screen that says "Your Access Token was successfully validated!".

Reviewer: ```suggestion
Now you will be able to run the application, browse to the "External API" page and press the "Call" button. The application will make a call to the external API endpoint and produce a message on the screen that says "Your Access Token was successfully validated!".
```
